File: dotnet/GameRL.Harmony/Bridge.cs
// IPC Bridge for communication with Rust harmony-server
// Acts as a SERVER - listens for connection from Rust side
// Protocol: length-prefixed JSON over Unix socket

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using GameRL.Harmony.Protocol;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace GameRL.Harmony
{
    /// <summary>
    /// IPC bridge that LISTENS for connections from Rust harmony-server.
    /// Thread-safe: receive runs on background thread, commands queued for main thread.
    /// </summary>
    public class Bridge : IDisposable
    {
        private Socket? _listenSocket;
        private Socket? _clientSocket;
        private NetworkStream? _stream;
        private Thread? _listenThread;
        private Thread? _receiveThread;
        private readonly ConcurrentQueue<GameMessage> _incomingQueue;
        private volatile bool _running;
        private readonly object _sendLock = new();
        private string? _socketPath;

        public event Action<RegisterAgentMessage>? OnRegisterAgent;
        public event Action<DeregisterAgentMessage>? OnDeregisterAgent;
        public event Action<ExecuteActionMessage>? OnExecuteAction;
        public event Action<ConfigureStreamsMessage>? OnConfigureStreams;
        public event Action<ResetMessage>? OnReset;
        public event Action<GetStateHashMessage>? OnGetStateHash;
        public event Action? OnShutdown;
        public event Action? OnClientConnected;

        public bool IsConnected => _running && _clientSocket?.Connected == true;
        public bool IsListening => _running && _listenSocket != null;

        public Bridge()
        {
            _incomingQueue = new ConcurrentQueue<GameMessage>();
        }

        /// <summary>
        /// Start listening for connections from Rust harmony-bridge
        /// </summary>
        public bool Listen(string socketPath)
        {
            try
            {
                _socketPath = socketPath;

                // Remove existing socket file if present
                if (File.Exists(socketPath))
                {
                    File.Delete(socketPath);
                }

                var endpoint = new UnixEndPoint(socketPath);
                _listenSocket = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
                _listenSocket.Bind(endpoint);
                _listenSocket.Listen(1);

                _running = true;
                _listenThread = new Thread(ListenLoop)
                {
                    IsBackground = true,
                    Name = "GameRL-Listen"
                };
                _listenThread.Start();

                Log($"Listening on {socketPath}");
                return true;
            }
            catch (Exception ex)
            {
                LogError($"Listen failed: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Background thread: wait for client connection
        /// </summary>
        private void ListenLoop()
        {
            while (_running && _listenSocket != null)
            {
                try
                {
                    Log("Waiting for Rust harmony-bridge connection...");
                    _clientSocket = _listenSocket.Accept();
                    _stream = new NetworkStream(_clientSocket, ownsSocket: false);

                    Log("Client connected!");
                    OnClientConnected?.Invoke();

                    // Start receive loop for this client
                    _receiveThread = new Thread(ReceiveLoop)
                    {
                        IsBackground = true,
                        Name = "GameRL-Receive"
                    };
                    _receiveThread.Start();

                    // Wait for receive thread to finish (client disconnect)
                    _receiveThread.Join();

                    // Clean up client connection
                    _stream?.Dispose();
                    _stream = null;
                    _clientSocket?.Dispose();
                    _clientSocket = null;

                    // Clear any stale messages from previous session
                    while (_incomingQueue.TryDequeue(out _)) { }

                    Log("Client disconnected, waiting for new connection...");
                }
                catch (SocketException ex) when (ex.SocketErrorCode == SocketError.OperationAborted)
                {
                    // Socket was closed, exit gracefully
                    break;
                }
                catch (Exception ex)
                {
                    if (_running)
                    {
                        LogError($"Listen error: {ex.Message}");
                    }
                }
            }
        }

        /// <summary>
        /// Background thread: continuously receive messages from connected client
        /// </summary>
        private void ReceiveLoop()
        {
            var lenBuffer = new byte[4];

            while (_running && _stream != null && _clientSocket?.Connected == true)
            {
                try
                {
                    // Read 4-byte length prefix (little-endian, matching Rust)
                    if (!ReadExact(lenBuffer, 0, 4))
                    {
                        Log("Connection closed by client");
                        break;
                    }

                    int length = BitConverter.ToInt32(lenBuffer, 0);
                    if (length <= 0 || length > 10_000_000)
                    {
                        LogError($"Invalid message length: {length}");
                        break;
                    }

                    // Read message body
                    var data = new byte[length];
                    if (!ReadExact(data, 0, length))
                    {
                        Log("Connection closed while reading message");
                        break;
                    }

                    // Deserialize and queue for main thread
                    var message = DeserializeMessage(data);
                    if (message != null)
                    {
                        _incomingQueue.Enqueue(message);
                    }
                }
                catch (Exception ex)
                {
                    if (_running)
                    {
                        LogError($"Receive error: {ex.Message}");
                    }
                    break;
                }
            }
        }

        private bool ReadExact(byte[] buffer, int offset, int count)
        {
            if (_stream == null) return false;

            int totalRead = 0;
            while (totalRead < count)
            {
                int read = _stream.Read(buffer, offset + totalRead, count - totalRead);
                if (read == 0) return false;
                totalRead += read;
            }
            return true;
        }

        /// <summary>
        /// Deserialize incoming JSON message based on "Type" field (PascalCase)
        /// </summary>
        private GameMessage? DeserializeMessage(byte[] data)
        {
            try
            {
                var json = Encoding.UTF8.GetString(data);

                // Diagnostic logging
                var preview = json.Length > 200 ? json.Substring(0, 200) + "..." : json;
                Log($"[C#←Rust] len={data.Length} json={preview}");

                var obj = JObject.Parse(json);
                var type = obj["Type"]?.ToString();

                if (string.IsNullOrEmpty(type))
                {
                    LogError("Message missing 'Type' field");
                    return null;
                }

                return type switch
                {
                    "RegisterAgent" => ParseRegisterAgent(obj),
                    "DeregisterAgent" => ParseDeregisterAgent(obj),
                    "ExecuteAction" => ParseExecuteAction(obj),
                    "ConfigureStreams" => ParseConfigureStreams(obj),
                    "Reset" => ParseReset(obj),
                    "GetStateHash" => new GetStateHashMessage(),
                    "Shutdown" => new ShutdownMessage(),
                    _ => null
                };
            }
            catch (Exception ex)
            {
                LogError($"Deserialize error: {ex.Message}");
                return null;
            }
        }

        private RegisterAgentMessage ParseRegisterAgent(JObject obj)
        {
            var msg = new RegisterAgentMessage
            {
                AgentId = obj["AgentId"]?.ToString() ?? "",
                AgentType = obj["AgentType"]?.ToString() ?? ""
            };

            var config = obj["Config"] as JObject;
            if (config != null)
            {
                msg.Config = new AgentConfig
                {
                    EntityId = config["EntityId"]?.ToString(),
                    ObservationProfile = config["ObservationProfile"]?.ToString() ?? "default"
                };
            }

            return msg;
        }

        private DeregisterAgentMessage ParseDeregisterAgent(JObject obj)
        {
            return new DeregisterAgentMessage
            {
                AgentId = obj["AgentId"]?.ToString() ?? ""
            };
        }

        private ExecuteActionMessage ParseExecuteAction(JObject obj)
        {
            // Convert JToken action to Dictionary for HarmonyRPC dispatch
            object? action = null;
            var actionToken = obj["Action"];
            if (actionToken != null && actionToken.Type != JTokenType.Null)
            {
                action = actionToken.ToObject<Dictionary<string, object>>();
            }

            return new ExecuteActionMessage
            {
                AgentId = obj["AgentId"]?.ToString() ?? "",
                Action = action,
                Ticks = obj["Ticks"]?.ToObject<uint>() ?? 1
            };
        }

        private ResetMessage ParseReset(JObject obj)
        {
            return new ResetMessage
            {
                Seed = obj["Seed"]?.ToObject<ulong?>(),
                Scenario = obj["Scenario"]?.ToString()
            };
        }

        private ConfigureStreamsMessage ParseConfigureStreams(JObject obj)
        {
            return new ConfigureStreamsMessage
            {
                AgentId = obj["AgentId"]?.ToString() ?? "",
                Profile = obj["Profile"]?.ToString() ?? "default"
            };
        }

        /// <summary>
        /// Process pending commands from the server. Call from main game thread.
        /// </summary>
        public void ProcessCommands()
        {
            while (_incomingQueue.TryDequeue(out var message))
            {
                try
                {
                    switch (message)
                    {
                        case RegisterAgentMessage m:
                            OnRegisterAgent?.Invoke(m);
                            break;
                        case DeregisterAgentMessage m:
                            OnDeregisterAgent?.Invoke(m);
                            break;
                        case ExecuteActionMessage m:
                            OnExecuteAction?.Invoke(m);
                            break;
                        case ConfigureStreamsMessage m:
                            OnConfigureStreams?.Invoke(m);
                            break;
                        case ResetMessage m:
                            OnReset?.Invoke(m);
                            break;
                        case GetStateHashMessage m:
                            OnGetStateHash?.Invoke(m);
                            break;
                        case ShutdownMessage:
                            OnShutdown?.Invoke();
                            break;
                    }
                }
                catch (Exception ex)
                {
                    LogError($"Handler error: {ex.Message}");
                }
            }
        }

        /// <summary>
        /// Send Ready message to Rust server
        /// </summary>
        public void SendReady(string name, string version, GameCapabilities capabilities)
        {
            Send(new ReadyMessage
            {
                Name = name,
                Version = version,
                Capabilities = capabilities
            });
        }

        /// <summary>
        /// Send agent registration response
        /// </summary>
        public void SendAgentRegistered(string agentId, object observationSpace, object actionSpace)
        {
            Send(new AgentRegisteredMessage
            {
                AgentId = agentId,
                ObservationSpace = observationSpace,
                ActionSpace = actionSpace
            });
        }

        /// <summary>
        /// Send step result with observation and reward
        /// </summary>
        public void SendStepResult(
            string agentId,
            object observation,
            double reward,
            Dictionary<string, double> rewardComponents,
            bool done,
            bool truncated,
            string? stateHash = null)
        {
            Send(new StepResultMessage
            {
                AgentId = agentId,
                Observation = observation,
                Reward = reward,
                RewardComponents = rewardComponents,
                Done = done,
                Truncated = truncated,
                StateHash = stateHash
            });
        }

        /// <summary>
        /// Send step results for multiple agents
        /// </summary>
        public void SendBatchStepResult(List<StepResultMessage> results)
        {
            Send(new BatchStepResultMessage
            {
                Results = results ?? new List<StepResultMessage>()
            });
        }

        /// <summary>
        /// Send reset complete with initial observation
        /// </summary>
        public void SendResetComplete(object observation, string? stateHash = null)
        {
            Send(new ResetCompleteMessage
            {
                Observation = observation,
                StateHash = stateHash
            });
        }

        /// <summary>
        /// Send error response
        /// </summary>
        public void SendError(int code, string message)
        {
            Send(new ErrorMessage
            {
                Code = code,
                Message = message
            });
        }

        /// <summary>
        /// Send state hash response
        /// </summary>
        public void SendStateHash(string hash)
        {
            Send(new StateHashMessage
            {
                Hash = hash
            });
        }

        /// <summary>
        /// Send vision stream configuration response
        /// </summary>
        public void SendStreamsConfigured(string agentId, List<Dictionary<string, object>> descriptors)
        {
            Send(new StreamsConfiguredMessage
            {
                AgentId = agentId,
                Descriptors = descriptors ?? new List<Dictionary<string, object>>()
            });
        }

        /// <summary>
        /// Send state update (async notification)
        /// </summary>
        public void SendStateUpdate(ulong tick, object state, List<GameEvent>? events = null)
        {
            Send(new StateUpdateMessage
            {
                Tick = tick,
                State = state,
                Events = events ?? new List<GameEvent>()
            });
        }

        private void Send(GameMessage message)
        {
            if (!_running || _stream == null || _clientSocket?.Connected != true) return;

            try
            {
                // Serialize message to JSON
                var json = SerializeMessage(message);
                var data = Encoding.UTF8.GetBytes(json);

                // Diagnostic logging
                var preview = json.Length > 200 ? json.Substring(0, 200) + "..." : json;
                Log($"[C#→Rust] len={data.Length} json={preview}");

                // Length prefix (4 bytes, little-endian to match Rust)
                var lenBytes = BitConverter.GetBytes(data.Length);

                lock (_sendLock)
                {
                    _stream.Write(lenBytes, 0, 4);
                    _stream.Write(data, 0, data.Length);
                    _stream.Flush();
                }
            }
            catch (Exception ex)
            {
                LogError($"Send error: {ex.Message}");
            }
        }

        /// <summary>
        /// Serialize message to JSON with "Type" field for Rust serde compatibility (PascalCase)
        /// </summary>
        private string SerializeMessage(GameMessage message)
        {
            var obj = new JObject();
            obj["Type"] = message.Type;

            switch (message)
            {
                case ReadyMessage m:
                    obj["Name"] = m.Name;
                    obj["Version"] = m.Version;
                    obj["Capabilities"] = JObject.FromObject(new
                    {
                        MultiAgent = m.Capabilities.MultiAgent,
                        MaxAgents = m.Capabilities.MaxAgents,
                        Deterministic = m.Capabilities.Deterministic,
                        Headless = m.Capabilities.Headless
                    });
                    break;

                case StepResultMessage m:
                {
                    var stepObj = SerializeStepResult(m);
                    foreach (var prop in stepObj)
                    {
                        obj[prop.Key] = prop.Value;
                    }
                    break;
                }

                case BatchStepResultMessage m:
                {
                    var results = new JArray();
                    foreach (var result in m.Results ?? new List<StepResultMessage>())
                    {
                        results.Add(SerializeStepResult(result));
                    }
                    obj["Results"] = results;
                    break;
                }

                case ResetCompleteMessage m:
                    obj["Observation"] = JToken.FromObject(m.Observation ?? new object());
                    if (m.StateHash != null)
                        obj["StateHash"] = m.StateHash;
                    break;

                case AgentRegisteredMessage m:
                    obj["AgentId"] = m.AgentId;
                    obj["ObservationSpace"] = JToken.FromObject(m.ObservationSpace ?? new object());
                    obj["ActionSpace"] = JToken.FromObject(m.ActionSpace ?? new object());
                    break;

                case StateHashMessage m:
                    obj["Hash"] = m.Hash;
                    break;

                case StreamsConfiguredMessage m:
                    obj["AgentId"] = m.AgentId;
                    obj["Descriptors"] = JToken.FromObject(m.Descriptors ?? new List<Dictionary<string, object>>());
                    break;

                case ErrorMessage m:
                    obj["Code"] = m.Code;
                    obj["Message"] = m.Message;
                    break;

                case StateUpdateMessage m:
                    obj["Tick"] = m.Tick;
                    obj["State"] = JToken.FromObject(m.State ?? new object());
                    obj["Events"] = JToken.FromObject(m.Events ?? new List<GameEvent>());
                    break;
            }

            return obj.ToString(Formatting.None);
        }

        private static JObject SerializeStepResult(StepResultMessage message)
        {
            var obj = new JObject
            {
                ["AgentId"] = message.AgentId,
                ["Observation"] = JToken.FromObject(message.Observation ?? new object()),
                ["Reward"] = message.Reward,
                ["RewardComponents"] = JToken.FromObject(message.RewardComponents ?? new Dictionary<string, double>()),
                ["Done"] = message.Done,
                ["Truncated"] = message.Truncated
            };

            if (message.StateHash != null)
            {
                obj["StateHash"] = message.StateHash;
            }

            return obj;
        }

        public void Dispose()
        {
            _running = false;

            // Close sockets to unblock threads
            _stream?.Dispose();
            _clientSocket?.Dispose();
            _listenSocket?.Dispose();

            // Clean up socket file
            if (_socketPath != null && File.Exists(_socketPath))
            {
                try { File.Delete(_socketPath); } catch { }
            }

            _listenThread?.Join(1000);
            _receiveThread?.Join(1000);
        }

        // Logging helpers - override in game-specific implementation
        protected virtual void Log(string message)
        {
            Console.WriteLine($"[GameRL] {message}");
        }

        protected virtual void LogError(string message)
        {
            Console.Error.WriteLine($"[GameRL] ERROR: {message}");
        }
    }
}

File: adapters/zomboid/mod/java/com/arkavo/gamerl/zomboid/Bridge.java
package com.arkavo.gamerl.zomboid;

import org.json.JSONObject;
import org.json.JSONArray;

import java.io.*;
import java.net.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Map;
import java.util.concurrent.*;
import java.util.function.Consumer;

/**
 * TCP server for communication with zomboid-bridge (Rust).
 * Uses length-prefixed JSON protocol matching game-bridge/src/protocol.rs
 */
public class Bridge {
    private final int port;
    private ServerSocket serverSocket;
    private Socket clientSocket;
    private DataInputStream input;
    private DataOutputStream output;
    private final ConcurrentLinkedQueue<JSONObject> incomingQueue;
    private Thread acceptThread;
    private Thread receiveThread;
    private volatile boolean running;

    // Event handlers
    private Consumer<Protocol.RegisterAgentMessage> onRegisterAgent;
    private Consumer<String> onDeregisterAgent;
    private Consumer<Protocol.ExecuteActionMessage> onExecuteAction;
    private Consumer<Protocol.ResetMessage> onReset;
    private Runnable onGetStateHash;
    private Consumer<Protocol.ConfigureStreamsMessage> onConfigureStreams;
    private Runnable onShutdown;

    public Bridge(int port) {
        this.port = port;
        this.incomingQueue = new ConcurrentLinkedQueue<>();
    }

    public boolean startServer() {
        try {
            serverSocket = new ServerSocket(port);
            serverSocket.setReuseAddress(true);
            running = true;

            acceptThread = new Thread(this::acceptLoop, "GameRL-Accept");
            acceptThread.setDaemon(true);
            acceptThread.start();

            return true;
        } catch (IOException e) {
            System.err.println("[GameRL] Failed to start server: " + e.getMessage());
            return false;
        }
    }

    public void stop() {
        running = false;
        cleanup();
        try {
            if (serverSocket != null) {
                serverSocket.close();
            }
        } catch (IOException e) {
            // ignore
        }
    }

    private void acceptLoop() {
        while (running) {
            try {
                System.out.println("[GameRL] Waiting for zomboid-bridge connection on port " + port + "...");
                clientSocket = serverSocket.accept();
                clientSocket.setTcpNoDelay(true); // Disable Nagle for low latency

                input = new DataInputStream(new BufferedInputStream(clientSocket.getInputStream()));
                output = new DataOutputStream(new BufferedOutputStream(clientSocket.getOutputStream()));

                System.out.println("[GameRL] Client connected!");
                GameRLMod.onClientConnected();

                receiveThread = new Thread(this::receiveLoop, "GameRL-Receive");
                receiveThread.setDaemon(true);
                receiveThread.start();
                receiveThread.join();

                cleanup();
            } catch (Exception e) {
                if (running) {
                    System.err.println("[GameRL] Accept error: " + e.getMessage());
                }
            }
        }
    }

    private void receiveLoop() {
        try {
            while (running && clientSocket != null && clientSocket.isConnected()) {
                // Read 4-byte length prefix (little-endian)
                byte[] lenBytes = new byte[4];
                input.readFully(lenBytes);
                int length = ByteBuffer.wrap(lenBytes).order(ByteOrder.LITTLE_ENDIAN).getInt();

                // Sanity check
                if (length > 64 * 1024 * 1024) {
                    System.err.println("[GameRL] Message too large: " + length);
                    break;
                }

                // Read message body
                byte[] data = new byte[length];
                input.readFully(data);

                String json = new String(data, "UTF-8");
                JSONObject msg = new JSONObject(json);
                incomingQueue.add(msg);
            }
        } catch (EOFException e) {
            System.out.println("[GameRL] Client disconnected");
        } catch (Exception e) {
            if (running) {
                System.err.println("[GameRL] Receive error: " + e.getMessage());
            }
        }
    }

    private void cleanup() {
        try {
            if (input != null) input.close();
            if (output != null) output.close();
            if (clientSocket != null) clientSocket.close();
        } catch (IOException e) {
            // ignore
        }
        input = null;
        output = null;
        clientSocket = null;
    }

    /**
     * Process queued messages - called from main game thread.
     */
    public void processIncoming() {
        JSONObject msg;
        while ((msg = incomingQueue.poll()) != null) {
            dispatchMessage(msg);
        }
    }

    private void dispatchMessage(JSONObject msg) {
        String type = msg.optString("Type");
        switch (type) {
            case "RegisterAgent":
                if (onRegisterAgent != null) {
                    onRegisterAgent.accept(Protocol.parseRegisterAgent(msg));
                }
                break;
            case "DeregisterAgent":
                if (onDeregisterAgent != null) {
                    onDeregisterAgent.accept(msg.optString("AgentId"));
                }
                break;
            case "ExecuteAction":
                if (onExecuteAction != null) {
                    onExecuteAction.accept(Protocol.parseExecuteAction(msg));
                }
                break;
            case "Reset":
                if (onReset != null) {
                    onReset.accept(Protocol.parseReset(msg));
                }
                break;
            case "GetStateHash":
                if (onGetStateHash != null) {
                    onGetStateHash.run();
                }
                break;
            case "ConfigureStreams":
                if (onConfigureStreams != null) {
                    onConfigureStreams.accept(Protocol.parseConfigureStreams(msg));
                }
                break;
            case "Shutdown":
                if (onShutdown != null) {
                    onShutdown.run();
                }
                break;
            default:
                System.err.println("[GameRL] Unknown message type: " + type);
        }
    }

    // === Send Methods ===

    private synchronized void send(JSONObject msg) {
        if (output == null) return;
        try {
            byte[] data = msg.toString().getBytes("UTF-8");
            byte[] lenBytes = ByteBuffer.allocate(4)
                .order(ByteOrder.LITTLE_ENDIAN)
                .putInt(data.length)
                .array();
            output.write(lenBytes);
            output.write(data);
            output.flush();
        } catch (IOException e) {
            System.err.println("[GameRL] Send error: " + e.getMessage());
        }
    }

    public void sendReady(String name, String version, boolean multiAgent, int maxAgents,
                          boolean deterministic, boolean headless) {
        JSONObject msg = new JSONObject();
        msg.put("Type", "Ready");
        msg.put("Name", name);
        msg.put("Version", version);

        JSONObject caps = new JSONObject();
        caps.put("MultiAgent", multiAgent);
        caps.put("MaxAgents", maxAgents);
        caps.put("Deterministic", deterministic);
        caps.put("Headless", headless);
        msg.put("Capabilities", caps);

        send(msg);
    }

    public void sendAgentRegistered(String agentId, Map<String, Object> observationSpace,
                                    Map<String, Object> actionSpace) {
        JSONObject msg = new JSONObject();
        msg.put("Type", "AgentRegistered");
        msg.put("AgentId", agentId);
        msg.put("ObservationSpace", new JSONObject(observationSpace));
        msg.put("ActionSpace", new JSONObject(actionSpace));
        send(msg);
    }

    public void sendStepResult(String agentId, Map<String, Object> observation,
                               double reward, Map<String, Double> rewardComponents,
                               boolean done, boolean truncated, String stateHash) {
        JSONObject msg = new JSONObject();
        msg.put("Type", "StepResult");
        msg.put("AgentId", agentId);
        msg.put("Observation", Protocol.observationToJson(observation));
        msg.put("Reward", reward);
        msg.put("RewardComponents", new JSONObject(rewardComponents));
        msg.put("Done", done);
        msg.put("Truncated", truncated);
        if (stateHash != null) {
            msg.put("StateHash", stateHash);
        }
        send(msg);
    }

    public void sendResetComplete(Map<String, Object> observation, String stateHash) {
        JSONObject msg = new JSONObject();
        msg.put("Type", "ResetComplete");
        msg.put("Observation", Protocol.observationToJson(observation));
        if (stateHash != null) {
            msg.put("StateHash", stateHash);
        }
        send(msg);
    }

    public void sendStateHash(String hash) {
        JSONObject msg = new JSONObject();
        msg.put("Type", "StateHash");
        msg.put("Hash", hash);
        send(msg);
    }

    public void sendStreamsConfigured(String agentId, JSONArray descriptors) {
        JSONObject msg = new JSONObject();
        msg.put("Type", "StreamsConfigured");
        msg.put("AgentId", agentId);
        msg.put("Descriptors", descriptors);
        send(msg);
    }

    public void sendError(int code, String message) {
        JSONObject msg = new JSONObject();
        msg.put("Type", "Error");
        msg.put("Code", code);
        msg.put("Message", message);
        send(msg);
    }

    // === Handler Setters ===

    public void setOnRegisterAgent(Consumer<Protocol.RegisterAgentMessage> handler) {
        this.onRegisterAgent = handler;
    }

    public void setOnDeregisterAgent(Consumer<String> handler) {
        this.onDeregisterAgent = handler;
    }

    public void setOnExecuteAction(Consumer<Protocol.ExecuteActionMessage> handler) {
        this.onExecuteAction = handler;
    }

    public void setOnReset(Consumer<Protocol.ResetMessage> handler) {
        this.onReset = handler;
    }

    public void setOnGetStateHash(Runnable handler) {
        this.onGetStateHash = handler;
    }

    public void setOnConfigureStreams(Consumer<Protocol.ConfigureStreamsMessage> handler) {
        this.onConfigureStreams = handler;
    }

    public void setOnShutdown(Runnable handler) {
        this.onShutdown = handler;
    }
}
